src/main/kotlin/graphics/scenery/net/NodeSubscriber.kt

```diff
@@ -1,117 +1,280 @@
 package graphics.scenery.net
 
-import com.esotericsoftware.kryo.Kryo
 import com.esotericsoftware.kryo.io.Input
-import com.jogamp.opengl.math.Quaternion
-import graphics.scenery.*
-import graphics.scenery.geometry.GeometryType
-import graphics.scenery.primitives.Arrow
-import graphics.scenery.primitives.Cylinder
-import graphics.scenery.primitives.Line
-import graphics.scenery.proteins.Protein
-import graphics.scenery.proteins.RibbonDiagram
+import graphics.scenery.Hub
+import graphics.scenery.Hubable
+import graphics.scenery.Node
+import graphics.scenery.Scene
 import graphics.scenery.utils.LazyLogger
-import graphics.scenery.utils.Statistics
-import graphics.scenery.volumes.TransferFunction
-import graphics.scenery.volumes.Volume
-import org.joml.Matrix4f
-import org.joml.Vector3f
-import org.objenesis.strategy.StdInstantiatorStrategy
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.zeromq.SocketType
 import org.zeromq.ZContext
 import org.zeromq.ZMQ
 import java.io.ByteArrayInputStream
-import java.io.StreamCorruptedException
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.LinkedBlockingQueue
+import kotlin.concurrent.thread
 
 
 /**
  * Created by ulrik on 4/4/2017.
  */
-class NodeSubscriber(override var hub: Hub?, val address: String = "tcp://localhost:6666", val context: ZContext = ZContext(4)) : Hubable {
-
+class NodeSubscriber(
+    override var hub: Hub?,
+    ip: String = "tcp://localhost",
+    portPublish: Int = 7777,
+    portBackchannel: Int = 6666,
+    val context: ZContext = ZContext(4),
+    startNetworkActivity: Boolean = true
+) : Hubable {
     private val logger by LazyLogger()
-    var nodes: ConcurrentHashMap<Int, Node> = ConcurrentHashMap()
-    var subscriber: ZMQ.Socket = context.createSocket(ZMQ.SUB)
     val kryo = NodePublisher.freeze()
 
+    private val addressSubscribe = "$ip:$portPublish"
+    private val addressBackchannel = "$ip:$portBackchannel"
+    var subscriber: ZMQ.Socket = context.createSocket(SocketType.SUB)
+    var backchannel: ZMQ.Socket = context.createSocket(SocketType.PUB)
+
+    private val networkObjects = hashMapOf<Int, NetworkWrapper<*>>()
+    private val eventQueue = LinkedBlockingQueue<NetworkEvent>()
+    private val waitingOnNetworkable = mutableMapOf<Int, List<Pair<NetworkEvent, WaitReason>>>()
+    private var listening = false
```

Comment:
This should probably be `@Synchronized` so it's the same on all threads.

Reply:
The hand-of point between the network thread and update/main-loop thread is the eventQueue which should be thread safe. The maps are ony accessed by the update thread and bools can't be @Synchronised